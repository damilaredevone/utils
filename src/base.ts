import { castArray } from './array'
import type { Fn } from './types'

/**
 * A function that takes in a value and a callback function. It then calls the callback function with the value and returns the value.
 *
 * @constant
 * @name tap
 * @kind variable
 * @type {<T>(value: T, callback?: ((value: T) => void) | undefined) => T}
 * @exports
 */
export const tap: <T>(
  value: T,
  callback?: ((value: T) => void) | undefined,
) => T = <T>(value: T, callback?: (value: T) => void): T => {
  if (isFunction<Fn>(callback)) {
    callback(value)
  }
  return value
}

/**
 * It's a function that takes in a function and returns the result of calling that function.
 *
 * @function
 * @name invoke
 * @kind variable
 * @param {Fn} fn
 * @returns {void}
 * @exports
 */
export const invoke = (fn: Fn): void => fn()

/**
 * It's a function that takes in an array of strings and returns a string.
 *
 * @function
 * @name classNames
 * @kind variable
 * @param {string[]} className
 * @returns {string}
 * @exports
 */
export const classNames = (className: string[]): string =>
  castArray<string>(className)?.filter(Boolean).join(' ')

/**
 * It's a type guard function that takes in a value and returns a boolean.
 *
 * @constant
 * @name notNullish
 * @kind variable
 * @type {<T>(v: T | null | undefined) => v is NonNullable<T>}
 * @exports
 */
export const notNullish: <T>(v: T | null | undefined) => v is NonNullable<T> = <
  T,
>(
  v: T | null | undefined,
): v is NonNullable<T> => v != null

/**
 * It's a type guard function that takes in a value and returns a boolean.
 *
 * @function
 * @name isObject
 * @kind variable
 * @param {unknown} value
 * @returns {value is object}
 * @exports
 */
export const isObject = (value: unknown): value is object =>
  Object.prototype.toString.call(value) === '[object Object]'

/**
 * It's a type guard function that takes in a value and returns a boolean.
 *
 * @function
 * @name isBoolean
 * @kind variable
 * @param {any} val
 * @returns {val is boolean}
 * @exports
 */
export const isBoolean = (val: any): val is boolean => typeof val === 'boolean'

/**
 * It's a type guard function that takes in a value and returns a boolean.
 *
 * @function
 * @name isFunction
 * @kind variable
 * @param {any} val
 * @returns {<T extends Function>(val: any) => val is T}
 * @exports
 */
export const isFunction = <T extends Function>(val: any): val is T =>
  typeof val === 'function'

/**
 * It's a function that scrolls to the top of the page.
 *
 * @function
 * @name scrollTop
 * @kind variable
 * @returns {void}
 * @exports
 */
export const scrollTop = (): void =>
  window.scrollTo({ top: 0, behavior: 'smooth' })

/**
 * It's a type guard function that takes in a value and returns a boolean.
 *
 * @constant
 * @name isKeyOf
 * @kind variable
 * @type {<T extends object>(obj: T, k: keyof any) => k is keyof T}
 * @exports
 */
export const isKeyOf = <T extends object>(
  obj: T,
  k: keyof any,
): k is keyof T => {
  return k in obj
}

/**
 * It's a function that takes in a value and returns a boolean.
 *
 * @function
 * @name value
 * @kind variable
 * @param {unknown} value
 * @returns {value is unknown}
 * @exports
 */
export const value = (value: unknown): value is unknown =>
  typeof value === 'function' ? value() : value

/**
 * It's a function that takes in a string and returns a string.
 *
 * @function
 * @name toBase64
 * @kind variable
 * @param {string} str
 * @returns {string}
 * @exports
 */
export const toBase64 = (str: string): string =>
  typeof window === 'undefined'
    ? Buffer.from(str).toString('base64')
    : window.btoa(str)

/**
 * It's a function that takes in a source object and an array of keys. It then returns a new object with the keys from the source object.
 *
 * @function
 * @name objectPick
 * @kind variable
 * @param {Record<string} source
 * @param {keys} any>
 * @param {any} string[]
 * @returns {Record<string, any>}
 * @exports
 */
export const objectPick = (
  source: Record<string, any>,
  keys: string[],
): Record<string, any> => {
  const obj: Record<string, any> = {}
  for (const key of keys) obj[key] = source[key]
  return obj
}

/**
 * A function that takes in a number and returns a string.
 *
 * @function
 * @name humanFileSize
 * @kind variable
 * @param {number} size
 * @returns {string}
 * @exports
 */
export const humanFileSize = (size: number): string => {
  const i = Math.floor(Math.log(size) / Math.log(1024))
  const v = size / 1024 ** i

  return `${v.toFixed(2)} ${['B', 'kB', 'MB', 'GB', 'TB'][i]}`
}
