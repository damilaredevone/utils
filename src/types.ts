export type Awaitable<T> = T | PromiseLike<T>;
export type Nullable<T> = T | null | undefined;
export type Arrayable<T> = T | Array<T>;

export type Fn<T = void, D = any> = (value?: any, defaultValue?: D) => T;
