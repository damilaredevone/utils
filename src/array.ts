import type { Nullable, Arrayable } from './types'

/**
 * A type guard.
 *
 * @constant
 * @name castArray
 * @kind variable
 * @type {<T>(array?: Nullable<Arrayable<T>>) => T[]}
 * @exports
 */
export const castArray = <T>(array?: Nullable<Arrayable<T>>): Array<T> => {
  array = array || []
  if (Array.isArray(array)) return array
  return [array]
}

/**
 * A type guard.
 *
 * @constant
 * @name flattenArrayable
 * @kind variable
 * @type {<T>(array?: Nullable<Arrayable<T | T[]>>) => T[]}
 * @exports
 */
export const flattenArrayable = <T>(
  array?: Nullable<Arrayable<T | Array<T>>>,
): Array<T> => castArray(array).flat(1) as Array<T>

/**
 * Returns the last element in an array.
 *
 * @constant
 * @name last
 * @kind variable
 * @type {<T>(array: readonly T[]) => T | undefined}
 * @exports
 */
export const last = <T>(array: readonly T[]): T | undefined =>
  position(array, -1)

/**
 * Returns the first element in an array.
 *
 * @constant
 * @name first
 * @kind variable
 * @type {<T>(array: readonly T[]) => T | undefined}
 * @exports
 */
export const first = <T>(array: readonly T[]): T | undefined =>
  position(array, 0)

/**
 * Returns an element at a certian position using the index.
 *
 * @constant
 * @name position
 * @kind variable
 * @type {<T>(array: [] | readonly T[], index: number) => T | undefined}
 * @exports
 */
export const position = <T>(
  array: readonly T[] | [],
  index: number,
): T | undefined => {
  const len = array.length
  if (!len) return undefined

  if (index < 0) index += len

  return array[index]
}

/**
 * A type guard.
 *
 * @constant
 * @name arrayIncludes
 * @kind variable
 * @type {<T>(array: T[] | readonly T[], item: unknown, fromIndex?: number) => item is T}
 * @exports
 */
export const arrayIncludes = <T>(
  array: T[] | readonly T[],
  item: unknown,
  fromIndex?: number,
): item is T => {
  return array.includes(item as T, fromIndex)
}

/**
 * A function that takes three arguments, `array`, `fromIndex`, and `toIndex`. It returns `void`.
 *
 * @constant
 * @name arrayMoveMutable
 * @kind variable
 * @type {<T>(array: Arrayable<T>[], fromIndex: number, toIndex: number) => void}
 * @exports
 */
export const arrayMoveMutable = <T>(
  array: Arrayable<T>[],
  fromIndex: number,
  toIndex: number,
): void => {
  const startIndex = fromIndex < 0 ? array.length + fromIndex : fromIndex

  if (startIndex >= 0 && startIndex < array.length) {
    const endIndex = toIndex < 0 ? array.length + toIndex : toIndex

    const [item] = array.splice(fromIndex, 1)
    array.splice(endIndex, 0, item)
  }
}

/**
 * A function that takes two arguments, `keys` and `obj`. It returns a new object that is a copy of `obj` but without the keys in `keys`.
 *
 * @constant
 * @name arrayMoveImmutable
 * @kind variable
 * @type {<T>(array: Arrayable<T>[], fromIndex: number, toIndex: number) => Arrayable<T>[]}
 * @exports
 */
export const arrayMoveImmutable = <T>(
  array: Arrayable<T>[],
  fromIndex: number,
  toIndex: number,
): Arrayable<T>[] => {
  const newArray = [...array]
  arrayMoveMutable(newArray, fromIndex, toIndex)

  return newArray
}

/**
 * A function that takes two arguments, `keys` and `obj`. It returns a new object that is a copy of `obj` but without the keys in `keys`.
 *
 * @constant
 * @name omit
 * @kind variable
 * @type {<T>(keys: T[], obj: Record<string, unknown>) => { [k: string]: unknown; }}
 * @exports
 */
export const omit = <T>(
  keys: Array<T>,
  obj: Record<string, unknown>,
): { [k: string]: unknown } =>
  Object.fromEntries(
    Object.entries(obj).filter(([k]: any) => !keys.includes(k)),
  )

export const uniq = <T extends any[]>(a: T): T[] => {
  return Array.from(new Set(a))
}
