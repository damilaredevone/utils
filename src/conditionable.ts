import type { Fn } from './types';

export const when = (
 value: boolean,
 callback?: Fn,
 defaultCallback?: Fn,
): void => {
 if (value) {
  return callback?.(this, value) ?? this;
 } else if (defaultCallback) {
  return defaultCallback?.(this, value);
 }

 return this;
};

export const unless = (
 value: boolean,
 callback?: Fn,
 defaultCallback?: Fn,
): void => {
 if (!value) {
  return callback?.(this, value) ?? this;
 } else if (defaultCallback) {
  return defaultCallback?.(this, value);
 }

 return this;
};
