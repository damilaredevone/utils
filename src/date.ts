export const formatDate = (date: string): string => {
  const monthNames = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ]
  const dateObj = new Date(date)
  const month = monthNames[dateObj.getMonth()]
  const day = String(dateObj.getDate()).padStart(2, '0')
  const year = dateObj.getFullYear()

  return `${day} ${month}, ${year}`
}

export const timestamp = (): number => Date.now()

export const getDaysInMonth = (
  month: number = new Date().getMonth(),
  year: number = new Date().getFullYear(),
): Date[] => {
  const date = new Date(year, month, 1)
  const days = [] as Date[]
  while (date.getMonth() === month) {
    days.push(new Date(date))
    date.setDate(date.getDate() + 1)
  }
  return days
}
