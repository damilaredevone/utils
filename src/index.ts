export * from './base';
export * from './array';
export * from './number';
export * from './str';
