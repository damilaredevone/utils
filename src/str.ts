export const capitalize = (str: string): string => str.toUpperCase();

export const sentenceCase = (string: string): string =>
 string?.replace(/(^(\w)|\s(\w))/g, (x) => x.toUpperCase());

export const truncate = (string: string, length = 30): string => {
 if (!string) {
  return string;
 }
 if (string.toString().length > length) {
  return `${string.toString().substring(0, length)}......`;
 }
 return string;
};
