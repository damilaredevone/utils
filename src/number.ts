import { flattenArrayable } from './array';

export const roundTo = (n: number, digits = 0): any => {
 let negative = false;
 let number: any = n;
 if (number < 0) {
  negative = true;
  number *= -1;
 }
 const multiplicator = 10 ** digits;
 number = parseFloat((number * multiplicator).toFixed(11));
 number = (Math.round(number) / multiplicator).toFixed(2);
 if (negative) {
  number = (number * -1).toFixed(2);
 }

 return number;
};

export const formatNumber = (n: number): string => {
 const ranges = [
  { divider: 1e18, suffix: 'E' },
  { divider: 1e15, suffix: 'P' },
  { divider: 1e12, suffix: 'T' },
  { divider: 1e9, suffix: 'G' },
  { divider: 1e6, suffix: 'M' },
  { divider: 1e3, suffix: 'k' },
 ];

 for (const range of ranges) {
  if (n >= range.divider) {
   return (n / range.divider).toFixed(1).toString() + range.suffix;
  }
 }

 return n.toString();
};

export const sum = (...args: number[] | number[][]) => {
 return flattenArrayable(args).reduce((a, b) => a + b, 0);
};

export const percentage = (value: string): string =>
 `${parseFloat(value).toFixed(2)}%`;
