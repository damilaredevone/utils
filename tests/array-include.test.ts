import { arrayIncludes } from '../src'

describe('ArrayIncludes', () => {
  it('returns true if item is included in array', () => {
    const array = [1, 2, 3]
    const item = 2
    const fromIndex = 0
    expect(arrayIncludes(array, item, fromIndex)).toBe(true)
    expect(arrayIncludes(array, item, fromIndex)).toMatchSnapshot()
  })

  it('returns false if item is not included in array', () => {
    const array = [1, 2, 3]
    const item = 4
    const fromIndex = 0
    expect(arrayIncludes(array, item, fromIndex)).toBe(false)
    expect(arrayIncludes(array, item, fromIndex)).toMatchSnapshot()
  })

  it('returns true if item is included in array starting from a specific index', () => {
    const array = [1, 2, 3, 2]
    const item = 2
    const fromIndex = 2
    expect(arrayIncludes(array, item, fromIndex)).toBe(true)
    expect(arrayIncludes(array, item, fromIndex)).toMatchSnapshot()
  })

  it('returns false if item is not included in array starting from a specific index', () => {
    const array = [1, 2, 3, 2]
    const item = 2
    const fromIndex = 4
    expect(arrayIncludes(array, item, fromIndex)).toBe(false)
    expect(arrayIncludes(array, item, fromIndex)).toMatchSnapshot()
  })
})
