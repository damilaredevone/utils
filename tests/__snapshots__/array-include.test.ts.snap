// Vitest Snapshot v1

exports[`ArrayIncludes > returns false if item is not included in array 1`] = `false`;

exports[`ArrayIncludes > returns false if item is not included in array starting from a specific index 1`] = `false`;

exports[`ArrayIncludes > returns true if item is included in array 1`] = `true`;

exports[`ArrayIncludes > returns true if item is included in array starting from a specific index 1`] = `true`;
